### PR TITLE
Feature/#15

### PR DESCRIPTION
**Pull request by chromatices**
* ap2 함수에서 torch.no_grad 테스트 진행했습니다.
* bit_shift_tensor 함수 내용중 for문을 np.where로 간략화 했습니다

### CODE REVIEW

File: utils/bit_shift_tensor.py [hunk 1]
```diff
@@ -3,43 +3,35 @@
...
     tensor_np = tensor_np.astype(np.int64)  # reformat cuz bit shift
 
-    flat_tensor = tensor_np.reshape(-1)  # reshape for every various tensor. Some has 2 by 2 but
+    flat_tensor = tensor_np.reshape(-1)
+    flat_tensor = flat_tensor.astype(np.int64)
+    # reshape for every various tensor. Some has 2 by 2 but
     # all tensor does not have 2 by 2
 
-    flat_bits = np.zeros(flat_tensor.shape)  # make numpy array
+    flat_bits = np.where(flat_tensor < 0, -1 * np.log2(1.0 + abs(flat_tensor)), np.log2(1.0 + flat_tensor))
```
Comment (ssaru) [inline utils/bit_shift_tensor.py]: 함수 분기 안하십니까?
제 말을 이해 못하셨나?...

File: utils/ap2.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import torch
+
+
+def approximate_power_of_two(x: torch.tensor):
+    with torch.no_grad():
+        return torch.sign(x) * 2 ** torch.round(torch.log2(torch.abs(x)))
```
Comment (ssaru) [inline utils/ap2.py]: `torch.no_grad()` 벤치마크한 결과 어디있나요?

File: utils/bit_shift_tensor.py (hunk 1 above)
Comment (chromatices) [reply utils/bit_shift_tensor.py]: 다시 수정하겠습니다.

File: utils/ap2.py (hunk 2 above)
Comment (chromatices) [reply utils/ap2.py]:
```python
>>> a = torch.randn(3, 3, requires_grad=True)
>>> b = a * a

print("tensor a : \n", a)

>>> tensor a : 
    tensor([[ 0.4220,  0.2116,  0.5364],
           [ 2.2112,  0.4657,  2.2447],
           [-1.1772, -0.1801,  0.9530]], requires_grad=True)

>>> print("\na.requres_grad : ",a.requires_grad)
   a.requres_grad :  True

>>> print("\na.grad : ",a.grad)
   a.grad :  None

>>> print("\na.grad_fn : ",a.grad_fn)
   a.grad_fn :  None

>>> print("\n a * a : \n", b)
    a * a : 
    tensor([[0.1781, 0.0448, 0.2877],
           [4.8894, 0.2168, 5.0389],
           [1.3857, 0.0324, 0.9082]], grad_fn=<MulBackward0>)

>>> print("\n(a*a).grad : ", b.grad)
   (a*a).grad :  None

>>> print("\n(a*a).grad_fn : ",b.grad_fn)
   (a*a).grad_fn :  <MulBackward0 object at 0x0000000016429C88>

>>> print("approximate_power_of_two(a).requires_grad :    ",approximate_power_of_two(a).requires_grad)
   approximate_power_of_two(a).requires_grad :  False

>>> print("approximate_power_of_two(a).grad : ",approximate_power_of_two(a).grad)
   approximate_power_of_two(a).grad :  None

>>> print("approximate_power_of_two(a).grad_fn : ",approximate_power_of_two(a).grad_fn)
   approximate_power_of_two(a).grad_fn :  None
```

File: utils/bit_shift_tensor.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+from ap2 import approximate_power_of_two
+import torch
+import numpy as np
+
+
+def extract_movement(bits: np.ndarray) -> np.ndarray:
+    movement = np.where(bits < 0, -1 * np.log(1.0 + abs(bits)), np.log2(1.0 + bits))
+    return movement
```
Comment (chromatices) [inline utils/bit_shift_tensor.py:8]: extract_movement라는 함수로 분기했습니다

File: utils/ap2.py (hunk 2 above)
Comment (ssaru) [reply utils/ap2.py]: 제가 말씀드린 것에 대한 의도를 전혀 이해하고 못하는 것 처럼 보이십니다.

왜 벤치마크시 `a*a` 연산을 사용하셨나요?
벤치마크를 하는 이유는 무엇인가요?
벤치마크를 통해서 무엇을 확인하고 싶은건가요?

1. 초기 구현하신 AP2 함수에는 `with torch.no_grad()`옵션이 구현되어있지 않았습니다.

2. 저는 `with torch.no_grad()`를 사용하지 않은 경우 코드 레벨에서 부작용이 발생할 가능성이 있다고 판단했습니다. 그래서 `with torch.no_grad()`옵션을 붙이라고 조언을 드렸던 것입니다.

3. 종환님은 단순히 제가 `with torch.no_grad()`를 붙이는게 좋다고 말씀드려서 `with torch.no_grad()`옵션을 붙이신 것으로 보였습니다. 이렇게 하면 제가 조언 드리는 이유가 뭐가 되나요? 그냥 아 그렇구나 하고 넘어가나요? 제가 틀렸으면 어떻게 하실겁니까?

4. 이러한 이유 때문에 AP2함수에 `with torch.no_grad()`가 있을 때와 없을 때, input으로 들어간 `torch.tensor`와 output `torch.tensor`의 멤버 변수인 `torch.tensor.grad` `torch.tensor.grad_fn`의 값이 어떻게 변화하는지 확인보라고 말씀드렸던 것입니다.

5. 이를 확인하여 AP2함수에서 `with torch.no_grad()`가 있을 때, 없을 때 어떠한 결과가 나오는지 확인해보고 `with torch.no_grad()`를 사용하고 사용하지 않은 모델을 `forward`, `backward`했을 때 어떠한 영향이 올지 생각해보라고 말씀드린 것입니다.

`torch.tensor.grad`와 `torch.tensor.grad_fn`의 역활이 무엇입니까?
그 둘은 PyTorch에서 어떻게 사용되나요?

File: utils/bit_shift_tensor.py (hunk 3 above)
Comment (ssaru) [reply utils/bit_shift_tensor.py:8]: 잘하셨습니다. 👏
제가 위에 말씀드린 벤치마크 결과까지만 보고 머지합시다.

**Review by ssaru — CHANGES_REQUESTED** [review]

File: utils/ap2.py (hunk 2 above)
Comment (chromatices) [reply utils/ap2.py]: ```a*a``` 연산을 한 이유는 torch tutorial에서 사용자가 만든 tensor의 경우 grad_fn의 값이 none으로 나와있다고 해서 변형이 필요하다고 느꼈습니다.
```a*a```의경우 연산을 통해 만들어진 텐서이므로 연산을 통해 비교하여 결과를 확인하고자 사용했습니다. 

아래 grad와 grad_fn의 역할은 튜토리얼을 다시 읽고 작성했습니다. 기능에 대해 숙지하고 작성하겠습니다.

================================================================
* ```grad```는 tensor의 변화도를 담고있다고 해서 일반적인 텐서와 연산을 통해 생성된 텐서의 정보량 비교를 위해 사용했습니다. 
* ```grad_fn```는 연산을 통해 tensor를 생성했을 경우 해당하는 연산을 보여주는 역할을 합니다.

================================================================

말씀해주신 벤치마크하고 전혀다른 결과를 배출했습니다. 이에 따라서 새로운 벤치마크를 생성했습니다. 검토 후 이상없으면 다시 PR 날리겠습니다.

```python
>>> a = torch.randn(3, 3, requires_grad=True)
>>> b = a * a
>>> print("tensor a : \n", a)
tensor a : 
 tensor([[ 1.2211,  1.7142, -1.4227],
        [-0.3912, -0.5827, -0.7742],
        [ 0.5215, -0.3521, -0.8508]], requires_grad=True)

>>> print("\na.requres_grad : ",a.requires_grad)
a.requres_grad :  True

>>> print("a.grad : ",a.grad)
a.grad :  None

>>> print("a.grad_fn : ",a.grad_fn)
a.grad_fn :  None

>>> print("\n a * a : ", b)
 a * a :  tensor([[1.4911, 2.9384, 2.0242],
        [0.1530, 0.3396, 0.5995],
        [0.2720, 0.1240, 0.7239]], grad_fn=<MulBackward0>)

>>> print("\na*a.requres_grad : ", b.requires_grad)
a*a.requres_grad :  True

>>> print("(a*a).grad : ", b.grad)
(a*a).grad :  None

>>> print("(a*a).grad_fn : ",b.grad_fn)
(a*a).grad_fn :  <MulBackward0 object at 0x0000000014E74948>

>>> print("\nAP2(a).requires_grad : ", AP2(a).requires_grad)
AP2(a).requires_grad :  True

>>> print("AP2(a).grad : ", AP2(a).grad)
AP2(a).grad :  None

>>> print("AP2(a).grad_fn :", AP2(a).grad_fn)
AP2(a).grad_fn : <MulBackward0 object at 0x0000000014E748C8>

>>> print("\napproximate_power_of_two(a).requires_grad : ",approximate_power_of_two(a).requires_grad)
approximate_power_of_two(a).requires_grad :  False

>>> print("approximate_power_of_two(a).grad : ",approximate_power_of_two(a).grad)
approximate_power_of_two(a).grad :  None

>>> print("approximate_power_of_two(a).grad_fn : ",approximate_power_of_two(a).grad_fn)
approximate_power_of_two(a).grad_fn :  None

>>> print("\nAP2(b).requires_grad : ", AP2(b).requires_grad)
AP2(b).requires_grad :  True

>>> print("AP2(b).grad : ", AP2(b).grad)
AP2(b).grad :  None

>>> print("AP2(b).grad_fn :", AP2(b).grad_fn)
AP2(b).grad_fn : <MulBackward0 object at 0x0000000014E752C8>

>>> print("\napproximate_power_of_two(a*a).requires_grad : ",approximate_power_of_two(b).requires_grad)
approximate_power_of_two(a*a).requires_grad :  False

>>> print("approximate_power_of_two(a*a).grad : ",approximate_power_of_two(b).grad)
approximate_power_of_two(a*a).grad :  None

>>> print("approximate_power_of_two(a*a).grad_fn : ",approximate_power_of_two(b).grad_fn)
approximate_power_of_two(a*a).grad_fn :  None
```

------------------------------------------------------------------------------------------
위에 결과로 with torch.no_grad() 메서드를 사용하여 가중치와 연산정보를 초기화해줄 수 있음을 알 수 있습니다.

Comment (ssaru) [reply utils/ap2.py]: `grad`는 딥러닝 모델에서 Gradient
`grad_func`은 Activation의 derivate. 
그러니까 `torch.tensor.require_grad=true`면 `forward`할 때 
연산을 하게되면 `grad_func`을 생성하게 되고
`backward`함수를 호출하면 `grad_fn`을 기반으로 `torch.tensor.grad`에 값을 생성해 내겠죠

그리고 그 `torch.tensor.grad`값을 이용해서 `optimizer`가 파라미터 업데이트할 때, weight에 업데이트 해주고요.

제 말이 어떤 말인지 이해가십니까?

그랬을 때, AP2함수에 `torch.no_grad()`를 쓰고 안쓰고 하면 어떤 결과가 나오나요?

Comment (chromatices) [reply utils/ap2.py]: 네 이해했습니다. 결과는 다음과 같이 나옵니다.

```python
>>> a = torch.randn(3,3, requires_grad = True)
>>> a
tensor([[ 1.2211,  1.7142, -1.4227],
        [-0.3912, -0.5827, -0.7742],
        [ 0.5215, -0.3521, -0.8508]], requires_grad=True)

>>> AP2(a)
tensor([[ 1.0000,  2.0000, -2.0000],
        [-0.5000, -0.5000, -1.0000],
        [ 0.5000, -0.2500, -1.0000]], grad_fn=<MulBackward0>)

>>> AP2(a).grad_fn
<MulBackward0 object at 0x0000000014E84CC8>

>>> approximate_power_of_two(a)
tensor([[ 1.0000,  2.0000, -2.0000],
        [-0.5000, -0.5000, -1.0000],
        [ 0.5000, -0.2500, -1.0000]])

>>> print(approximate_power_of_two(a).grad_fn)
None
```

```torch.no_grad()```를 통해 derivate를 초기화해줌을 알 수 있습니다.

Comment (chromatices) [reply utils/ap2.py]: batchnorm 조사하면서 알게된건데, batchnorm을 적용하면 ```grad_fn```에
 ```grad_fn=<NativeBatchNormBackward>``` 이라는 옵션이 붙어서 향후에 AP2연산을 진행하면 제대로 학습이 되지 않을 우려가 있을것 같습니다.

```python
>>> a = torch.randn(2,2)
>>> test = nn.BatchNorm1d(2)
>>> result = test(a)

>>> print("a : ", a)
a :  tensor([[-0.5552,  0.9636],
        [ 0.6747,  0.4836]])

>>> print("batchnorm(a) : ", result)
batchnorm(a) :  tensor([[-0.7247,  0.5159],
        [ 0.7247, -0.5159]], grad_fn=<NativeBatchNormBackward>)

```

Comment (ssaru) [reply utils/ap2.py]: `torch.no_grad()`는 derivate를 초기화해주는 것이 아닙니다.
`torch.no_grad()`는 forward할 때, 연산의 drivate를 추적하지 않겠다는 의미입니다.

Comment (ssaru) [reply utils/ap2.py]: 훌륭하십니다. 저도 생각하고 있지 못했던 것인데!! ㅎㅎ
그럼 shift value를 이용해서 어떻게 backprop을 할 수 있을까요?

관련 연구가 있을까요?
찾아보시겠어요?

**Review by ssaru — APPROVED** [review]
